File: abrHelpers.py
# uses abr web services to find acn (if exists) based on abn, returns None if business doesn't have an acn
# def abrSearchABN(abn):
#     return ("Trustworthy subcontractor solutions", None)

# # uses abr web services to find abn and acn (if exists) based on a business name
# def abrSearchName(name):
#     return (59537738403, None)

import requests
import xml.etree.ElementTree as ET
import json
import urllib
import datetime

# GUID for accessing the web services
GUID = "d6c41993-5ce1-41cd-a671-a7249e243efb"
ABR_URL = "https://abr.business.gov.au/"
ELEM_PREFIX = "{http://abr.business.gov.au/ABRXMLSearchRPC/literalTypes}"
ACN_PREFIX = '{http://abr.business.gov.au/ABRXMLSearch/}'
''' 
Information provided by this path:
History of business/trading names (trading names are generally obsolete), people tied to this abn, asicNumber, state, Periods of Active/Cancelled, 
Available on this path but not returned by this function: Public/Private Company, Postcode, Individual/Company (I think(?) already determined by existance of acn or not)
'''
def abrSearchABN(abn):
    includeHistoricalDetails = "Y"
    path = f"abrxmlsearchRPC/AbrXmlSearch.asmx/SearchByABNv201205?searchString={abn}&includeHistoricalDetails={includeHistoricalDetails}&authenticationGuid={GUID}"
    outputDict = {}
    # try:
    response = requests.get(ABR_URL + path, timeout=10)
    response.raise_for_status()
    # print(response.content)

    data = ET.fromstring(response.content)[1].findall(f"{ELEM_PREFIX}businessEntity201205")[0]
    tradingNamesElem = data.findall(f"{ELEM_PREFIX}mainTradingName")
    businessNamesElem = data.findall(f"{ELEM_PREFIX}businessName")
    legalNamesElem = data.findall(f"{ELEM_PREFIX}legalName")
    asicElem = data.find(f"{ELEM_PREFIX}ASICNumber")
    addressElem = data.findall(f"{ELEM_PREFIX}mainBusinessPhysicalAddress")

    statusTimeline = []
    statusElem = data.findall(f"{ELEM_PREFIX}entityStatus")
    for status in statusElem:
        effFrom = status.find(f"{ELEM_PREFIX}effectiveFrom")
        effTo = status.find(f"{ELEM_PREFIX}effectiveTo")
        if effTo is None:
            date = (effFrom.text, "Present")
        else:
            date = (effFrom.text, effTo.text)
        statusTimeline.append((status.find(f"{ELEM_PREFIX}entityStatusCode").text, date))
        # print(status.find(f"{ELEM_PREFIX}entityStatusCode").text)

    # print(data)
    # for child in data:
    #     print(child.tag, child.attrib)


    tradingNames = []
    for name in tradingNamesElem:
        effFrom = name.find(f"{ELEM_PREFIX}effectiveFrom")
        effTo = name.find(f"{ELEM_PREFIX}effectiveTo")
        if effTo is None:
            date = (effFrom.text, "Present")
        else:
            date = (effFrom.text, effTo.text)
        
        orgName = name.find(f"{ELEM_PREFIX}organisationName").text
        tradingNames.append((orgName, date))
    
    businessNames = []
    for name in businessNamesElem:
        effFrom = name.find(f"{ELEM_PREFIX}effectiveFrom")
        effTo = name.find(f"{ELEM_PREFIX}effectiveTo")
        if effTo is None:
            date = (effFrom.text, "Present")
        else:
            date = (effFrom.text, effTo.text)
        
        orgName = name.find(f"{ELEM_PREFIX}organisationName").text
        businessNames.append((orgName, date))
    
    legalNames = []
    for name in legalNamesElem:
        effFrom = name.find(f"{ELEM_PREFIX}effectiveFrom")
        effTo = name.find(f"{ELEM_PREFIX}effectiveTo")
        if effTo is None:
            date = (effFrom.text, "Present")
        else:
            date = (effFrom.text, effTo.text)
        givenName = name.find(f"{ELEM_PREFIX}givenName").text
        otherGivenName = name.find(f"{ELEM_PREFIX}otherGivenName").text
        otherGivenName = ' ' + otherGivenName if otherGivenName else ''
        familyName = name.find(f"{ELEM_PREFIX}familyName").text
        personName = f'{givenName}{otherGivenName} {familyName}'
        legalNames.append((personName, date))
    
    states = set()
    for address in addressElem:
        # print(address.tag, address.attrib)

        states.add(address.find(f"{ELEM_PREFIX}stateCode").text)

    # print(asicElem.tag)
    acn = None
    if not asicElem is None:
        acn = asicElem.text
    outputDict["businessNames"] = businessNames
    outputDict["tradingNames"] = tradingNames
    outputDict["relavantPeople"] = legalNames
    outputDict["statusTimeline"] = statusTimeline
    outputDict["states"] = states
    outputDict["acn"] = acn
    return outputDict

    # except:
    #     print("oopies")
def abrSearchName(name, option="businessName"):
    """
    Uses ABR web services to find ABN based on a name.
    
    Args:
        name (str): query string
        option (str): "businessName" if query is the business name, "legalName" if query is for the registered person for the business
        
    Returns:
        abn (string): The Australian Business Number
    """
    
    path = "abrxmlsearchRPC/AbrXmlSearch.asmx/ABRSearchByNameAdvancedSimpleProtocol2017"

    business_name = "Y" if option == "businessName" else "N"
    legal_name = "Y" if option == "personName" else "N"
    states_query = "&NSW=&SA=&ACT=&VIC=&WA=&NT=&QLD=&TAS=" # change when scaling up
    query = f"?name={urllib.parse.quote_plus(name)}&postcode=&legalName={legal_name}&tradingName={business_name}&businessName={business_name}&activeABNsOnly=Y{states_query}&authenticationGuid={GUID}&searchWidth=&minimumScore=&maxSearchResults="
    
    response = requests.get(ABR_URL + path + query, timeout=10)
    print(ABR_URL + path + query)
    response.raise_for_status()
    # print(response.content)

    # data = ET.fromstring(response.content)[1].findall(f"response")[0]
    data = ET.fromstring(response.content)[1].findall(f"{ELEM_PREFIX}searchResultsList")[0]
    resultsElem = data.findall(f"{ELEM_PREFIX}searchResultsRecord")

    abns = {}
    for result in resultsElem:
        # result also contains postcode/state
        abn = result.find(f"{ELEM_PREFIX}ABN").find(f"{ELEM_PREFIX}identifierValue").text
        names_to_check = ["businessName", "legalName", "mainName", "mainTradingName", "otherTradingName"]
        score = None
        for name_type in names_to_check:
            element = result.find(f"{ELEM_PREFIX}{name_type}")
            # print(element)
            if not element is None:
                score = element.find(f"{ELEM_PREFIX}score").text
                break 
        if score is None:
            for child in result:
                print(child.tag)
        if not abn in abns:
            abns[abn] = int(score) # if error here check parent tag is included in names_to_check, it will be one of the printed tags in line 164
            
        # abns.append((abn, int(score))) # if error here check parent tag is included in names_to_check, it will be one of the printed tags in line 164
            
    return list(abns.items())
    



'https://abr.business.gov.au/abrxmlsearch/AbrXmlSearch.asmx/SearchByASICv201408?searchString=166+205+938&includeHistoricalDetails=Y&authenticationGuid=d6c41993-5ce1-41cd-a671-a7249e243efb'
def abrSearchACN(acn):
    path = 'abrxmlsearch/AbrXmlSearch.asmx/SearchByASICv201408'
    query = f'?searchString={acn}&includeHistoricalDetails=Y&authenticationGuid={GUID}'
    response = requests.get(ABR_URL + path + query, timeout=10)
    # print(ABR_URL + path + query)
    response.raise_for_status()
    

    data = ET.fromstring(response.content)[1].findall(f"{ACN_PREFIX}businessEntity201408")[0]
    abnElem = data.find(f"{ACN_PREFIX}ABN")

    for elem in abnElem:
        print(elem.tag)


    return

abrSearchACN(166205938)

File: trustworthy-frontend/src/lib/abrHelpers.ts
import { parseString } from 'xml2js';

// GUID for accessing the web services
const GUID = "d6c41993-5ce1-41cd-a671-a7249e243efb";
const ABR_URL = "https://abr.business.gov.au/abrxmlsearchRPC/AbrXmlSearch.asmx";

interface NameWithDate {
  name: string;
  startDate: string;
  endDate: string | null;
}

interface StatusTimeline {
  status: string;
  startDate: string;
  endDate: string | null;
}

export interface ABRSearchResult {
  abn: string;
  businessNames: NameWithDate[];
  tradingNames: NameWithDate[];
  relevantPeople: NameWithDate[];
  statusTimeline: StatusTimeline[];
  states: string[];
  acn: string | null;
}

export interface ABRNameSearchResult {
  abn: string;
  score: number;
}

/**
 * Information provided by this path:
 * History of business/trading names (trading names are generally obsolete), people tied to this abn, asicNumber, state, Periods of Active/Cancelled, 
 * Available on this path but not returned by this function: Public/Private Company, Postcode, Individual/Company (I think(?) already determined by existance of acn or not)
 */
export async function abrSearchABN(abn: string): Promise<ABRSearchResult | string> {
  const includeHistoricalDetails = "Y";
  const path = buildABRPath(abn, includeHistoricalDetails);

  try {
    const response = await fetchABRData(path);
    const xmlContent = await response.text();

    return parseABRResponse(xmlContent, abn);
  } catch (error) {
    console.error('Error in abrSearchABN:', error);
    return `Error with ${abn}`;
  }
}

/** Builds the full ABR API request path */
function buildABRPath(abn: string, includeHistoricalDetails: string): string {
  return `/SearchByABNv201205?searchString=${abn}&includeHistoricalDetails=${includeHistoricalDetails}&authenticationGuid=${GUID}`;
}

/** Executes the fetch call and throws on HTTP error */
async function fetchABRData(path: string): Promise<Response> {
  const response = await fetch(ABR_URL + path, {
    method: 'GET',
    headers: { 'Accept': 'application/xml' },
  });

  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }
  return response;
}

/** Parses the XML response and returns the structured ABRSearchResult */
function parseABRResponse(xmlContent: string, abn: string): Promise<ABRSearchResult> {
  return new Promise((resolve, reject) => {
    parseString(xmlContent, (err: any, result: any) => {
      if (err) {
        console.error('XML parsing error:', err);
        reject(`Error parsing XML for ${abn}`);
        return;
      }

      try {
        const businessEntity = extractBusinessEntity(result);
        const outputDict: ABRSearchResult = initializeABRResult(abn);

        parseTradingNames(businessEntity, outputDict);
        parseBusinessNames(businessEntity, outputDict);
        parseLegalNames(businessEntity, outputDict);
        parseEntityStatuses(businessEntity, outputDict);
        parseStates(businessEntity, outputDict);
        parseACN(businessEntity, outputDict);

        resolve(outputDict);
      } catch (parseError) {
        console.error('Error parsing business entity:', parseError);
        reject(`Error parsing response for ${abn}`);
      }
    });
  });
}

/** Navigates the XML to find the main businessEntity node */
function extractBusinessEntity(result: any): any {
  return result.ABRPayloadSearchResults.response[0].businessEntity201205[0]
}

/** Initializes an empty ABRSearchResult structure */
function initializeABRResult(abn: string): ABRSearchResult {
  return {
    abn,
    businessNames: [],
    tradingNames: [],
    relevantPeople: [],
    statusTimeline: [],
    states: [],
    acn: null
  };
}

function parseTradingNames(businessEntity: any, outputDict: ABRSearchResult) {
  const tradingNames = businessEntity['mainTradingName'] || [];
  for (const name of tradingNames) {
    const orgName = name['organisationName'][0];
    const effFrom = name['effectiveFrom'][0];
    const effTo = name['effectiveTo'] ? name['effectiveTo'][0] : null;

    outputDict.tradingNames.push({ name: orgName, startDate: effFrom, endDate: effTo });
  }
}

function parseBusinessNames(businessEntity: any, outputDict: ABRSearchResult) {
  const businessNames = businessEntity['businessName'] || [];
  for (const name of businessNames) {
    const orgName = name['organisationName'][0];
    const effFrom = name['effectiveFrom'][0];
    const effTo = name['effectiveTo'] ? name['effectiveTo'][0] : null;

    outputDict.businessNames.push({ name: orgName, startDate: effFrom, endDate: effTo });
  }
}

function parseLegalNames(businessEntity: any, outputDict: ABRSearchResult) {
  const legalNames = businessEntity['legalName'] || [];
  for (const name of legalNames) {
    const givenName = name['givenName'][0];
    const otherGivenName = name['otherGivenName'] ? name['otherGivenName'][0] : '';
    const familyName = name['familyName'][0];
    const personName = `${givenName}${otherGivenName ? ' ' + otherGivenName : ''} ${familyName}`;
    const effFrom = name['effectiveFrom'][0];
    const effTo = name['effectiveTo'] ? name['effectiveTo'][0] : null;

    outputDict.relevantPeople.push({ name: personName, startDate: effFrom, endDate: effTo });
  }
}

function parseEntityStatuses(businessEntity: any, outputDict: ABRSearchResult) {
  const entityStatuses = businessEntity['entityStatus'] || [];
  for (const status of entityStatuses) {
    const statusCode = status['entityStatusCode'][0];
    const effFrom = status['effectiveFrom'][0];
    const effTo = status['effectiveTo'] ? status['effectiveTo'][0] : null;

    outputDict.statusTimeline.push({ status: statusCode, startDate: effFrom, endDate: effTo });
  }
}

function parseStates(businessEntity: any, outputDict: ABRSearchResult) {
  const addresses = businessEntity['mainBusinessPhysicalAddress'] || [];
  const states = new Set<string>();
  for (const address of addresses) {
    if (address['stateCode']) {
      states.add(address['stateCode'][0]);
    }
  }
  outputDict.states = Array.from(states);
}

function parseACN(businessEntity: any, outputDict: ABRSearchResult) {
  if (businessEntity['ASICNumber']) {
    outputDict.acn = businessEntity['ASICNumber'][0];
  }
}
/**
 * Uses ABR web services to find ABN based on a name.
 * 
 * @param name - query string
 * @param option - "businessName" if query is the business name, "legalName" if query is for the registered person for the business
 * @returns Array of tuples with ABN and score
 */
export async function abrSearchName(name: string, option: string = "businessName"): Promise<ABRNameSearchResult[]> {
  const path = "/ABRSearchByNameAdvancedSimpleProtocol2017";
  
  const business_name = option === "businessName" ? "Y" : "N";
  const legal_name = option === "personName" ? "Y" : "N";
  const states_query = "&NSW=&SA=&ACT=&VIC=&WA=&NT=&QLD=&TAS="; // change when scaling up
  const encodedName = encodeURIComponent(name);
  const query = `?name=${encodedName}&postcode=&legalName=${legal_name}&tradingName=${business_name}&businessName=${business_name}&activeABNsOnly=Y${states_query}&authenticationGuid=${GUID}&searchWidth=&minimumScore=&maxSearchResults=`;
  
  try {
    const response = await fetch(ABR_URL + path + query, {
      method: 'GET',
      headers: {
        'Accept': 'application/xml',
      },
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const xmlContent = await response.text();
    // console.log(ABR_URL + path + query);

    return new Promise((resolve, reject) => {
      parseString(xmlContent, (err: any, result: any) => {
        if (err) {
          console.error('XML parsing error:', err);
          reject('Error parsing XML response');
          return;
        }

        try {
          const searchResults = result.ABRPayloadSearchResults.response[0].searchResultsList[0].searchResultsRecord || [];
          
          const abns: { [key: string]: number } = {};
          
          for (const result of searchResults) {
            const abn = result['ABN'][0]['identifierValue'][0];
            
            const names_to_check = ["businessName", "legalName", "mainName", "mainTradingName", "otherTradingName"];
            let score: number | null = null;
            
            for (const name_type of names_to_check) {
              const element = result[name_type];
              if (element && element[0] && element[0]['score']) {
                score = parseInt(element[0]['score'][0]);
                break;
              }
            }
            
            if (score === null) {
              console.log('Available elements:', Object.keys(result));
              continue;
            }
            
            if (!abns[abn]) {
              abns[abn] = score;
            }
          }
          
          const results: ABRNameSearchResult[] = Object.entries(abns).map(([abn, score]) => ({
            abn,
            score
          }));
          
          resolve(results);
        } catch (parseError) {
          console.error('Error parsing search results:', parseError);
          reject('Error parsing search results');
        }
      });
    });
  } catch (error) {
    console.error('Error in abrSearchName:', error);
    return [];
  }
}

export async function abrSearchACN(acn: string): Promise<string | null> {
  // Implementation for ACN search - placeholder for now
  // This would need to be implemented based on the ABR API documentation
  console.log(`ACN search not implemented yet for: ${acn}`);
  return null;
}
